### PR TITLE
ARROW-1135: [C++] Use clang 4.0 in one of the Linux builds

### PR DESCRIPTION
**Pull request by wesm**
This will help us keep the Linux build clean on latest clang

### CODE REVIEW

**Comment by cpcloud** [comment]
sweet, +1

**Comment by wesm** [comment]
While working on this I discovered we were indicating externalproject dependencies incorrectly, but I think it's fixed now (created a synthetic target to capture all of the build pre-requisites for the arrow library)

**Comment by wesm** [comment]
+1. I'm not sure why the Travis CI build didn't run, but it ran on my fork: https://travis-ci.org/wesm/arrow/builds/246412973

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 1]
```diff
@@ -367,12 +370,8 @@ if (ARROW_IPC)
     set(FLATBUFFERS_VENDORED 0)
   endif()
 
-  if(RAPIDJSON_VENDORED)
-    set(ARROW_DEPENDENCIES ${ARROW_DEPENDENCIES} rapidjson_ep)
-  endif()
-
   if(FLATBUFFERS_VENDORED)
-    set(ARROW_DEPENDENCIES ${ARROW_DEPENDENCIES} flatbuffers_ep)
```
Comment (wesm) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake]: For future reference, this way of declaring dependencies as a list with EP's seems to not work with all versions of CMake. It works with library targets though for some reason

**Pull request closed without merging**